librz/bin/format/elf/elf_dynamic.c

```diff
@@ -16,80 +16,105 @@ static bool get_dynamic_entry(ELFOBJ *bin, ut64 offset, Elf_(Dyn) * entry) {
 	return true;
 }
 
-static void fill_dt_dynamic(ELFOBJ *bin, RzBinElfDtDynamic *ptr, ut64 offset, ut64 size) {
+static bool get_dynamic_entry(ELFOBJ *bin, ut64 offset, Elf_(Dyn) * entry) {
+	if (!get_dynamic_entry_aux(bin, offset, entry)) {
+		RZ_LOG_WARN("Dynamic entry read failed.\n");
+		return false;
+	}
+
+	return true;
+}
+
+static bool fill_dt_dynamic(ELFOBJ *bin, RzBinElfDtDynamic *ptr, ut64 offset, ut64 size) {
 	for (ut64 entry_offset = offset; entry_offset < offset + size; entry_offset += sizeof(Elf_(Dyn))) {
 		Elf_(Dyn) entry;
-
 		if (!get_dynamic_entry(bin, entry_offset, &entry)) {
-			break;
+			return false;
 		}
 
 		if (entry.d_tag == DT_NULL) {
 			break;
 		}
 
 		if (entry.d_tag == DT_NEEDED) {
-			rz_vector_push(ptr->dt_needed, &entry.d_un.d_val);
+			if (!rz_vector_push(ptr->dt_needed, &entry.d_un.d_val)) {
+				RZ_LOG_WARN("Dynamic entry (DT_NEEDED) allocation failed.\n");
+				return false;
+			}
 		} else if (!ht_uu_insert(ptr->info, entry.d_tag, entry.d_un.d_val)) {
 			RZ_LOG_WARN("Dynamic tag %" PFMT64d " already handled\n", (ut64)entry.d_tag);
 		}
 	}
+
+	return true;
 }
 
 static bool init_dt_dynamic(ELFOBJ *bin, RzBinElfDtDynamic *ptr) {
 	RzBinElfSegment *segment = Elf_(rz_bin_elf_get_segment_with_type)(bin, PT_DYNAMIC);
 	if (!segment) {
+		RZ_LOG_WARN("No DT_DYNAMIC entry detected.\n")
 		return false;
 	}
 
-	ut64 size = segment->data.p_filesz;
-	ut64 offset = Elf_(rz_bin_elf_v2p_new)(bin, segment->data.p_vaddr);
-	if (offset == UT64_MAX || !size || offset + size > bin->size) {
+	if (!segment->is_valid) {
+		RZ_LOG_WARN("Invalid DT_DYNAMIC segment.\n")
 		return false;
 	}
 
-	fill_dt_dynamic(bin, ptr, offset, size);
+	ut64 offset = Elf_(rz_bin_elf_v2p_new)(bin, segment->data.p_vaddr);
+	if (offset == UT64_MAX) {
+		RZ_LOG_WARN("Segment offset conversion failed.\n")
+		return false;
+	}
 
-	return true;
+	return fill_dt_dynamic(bin, ptr, offset, segment->data.p_filesz);
 }
 
 RZ_BORROW RzVector *Elf_(rz_bin_elf_get_dt_needed)(RZ_NONNULL ELFOBJ *bin) {
-	rz_return_val_if_fail(bin && bin->dt_dynamic && bin->dt_dynamic->dt_needed, NULL);
+	rz_return_val_if_fail(bin, NULL);
+
+	if (!Elf_(rz_bin_elf_has_dt_dynamic)(bin)) {
+		return NULL;
+	}
 
 	return bin->dt_dynamic->dt_needed;
 }
 
 RZ_OWN RzBinElfDtDynamic *Elf_(rz_bin_elf_dt_dynamic_new)(RZ_NONNULL ELFOBJ *bin) {
-	RzBinElfDtDynamic *result = RZ_NEW(RzBinElfDtDynamic);
+	RzBinElfDtDynamic *result = RZ_NEW0(RzBinElfDtDynamic);
 	if (!result) {
+		RZ_LOG_WARN("Dynamic section allocation failed.\n");
 		return NULL;
 	}
 
 	result->info = ht_uu_new0();
 	if (!result->info) {
-		free(result);
+		RZ_LOG_WARN("Dynamic section allocation failed.\n");
+		Elf_(rz_bin_elf_dt_dynamic_free)(result);
 		return NULL;
 	}
 
 	result->dt_needed = rz_vector_new(sizeof(Elf_(Word)), NULL, NULL);
 	if (!result->dt_needed) {
-		ht_uu_free(result->info);
-		free(result);
+		RZ_LOG_WARN("Dynamic section allocation failed.\n");
```

Comment:
here use RZ_LOG_ERROR. if it is a fail, then is ERROR, like for example when you fail to alloc.
When is a problem, like you expect to have 255 max values but you have 1000, then is a warning.

Reply:
same for all failed allocs in this code.

Reply:
Ty